llvm/test/Instrumentation/AddressSanitizer/asan-pass-second-run.ll

```diff
@@ -0,0 +1,56 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --check-globals all --version 5
+; This test checks in the second run, function is not instrumented again.
+; RUN: opt < %s -passes=asan,asan -S | FileCheck %s
```

Comment:
Why is this needed?

Time to time we had discussion to do exactly like that, but usually it's not hard to make sure  that pass is unique in pipeline 

There is also ConstructorKind::kNone when constructor is not inserted.

Inconsistency with other sanitizers is also not nice.

Reply:
We found this issue when a HIP program is first compiled to bitcode with -fsanitize=address, then the bitcode is compiled to object file with -fsanitize=address again. This lead to calculation of shadow memory twice due to double instrumentation.

As you pointed out, ConstructorKind::None can lead to asan.module_ctor to be not present in the module. Do you suggest any other way to check if module is instrumented?

Reply:
@vitalybuka  I think it's a good idea to make all sanitizers idempotent.  Your guidance on preferred approaches would be appreciated.  With separate compilation, runtime compilation, etc. it seems hard to always "make sure that pass is unique in the pipeline".

Reply:
It would be nice to have some generic approach, which will work regardless of constructors.

How about about Module::addModuleFlag("nosanitize") which will be set and checked by every sanitizer?

I am not sure if this require RFC on discourse. @nikic @efriedma-quic WDYT?

Reply:
Have updated PR to use "nosanitize" module flag. Could also check the return value for key "nosanitize". Currently its set to "1" at end of asan pass run.

Reply:
Have added checks to identify if "nosanitize" module flag is present before instrumentation for hwsan, msan and dfsan.

Reply:
I'm not sure a generic nosanitize module flag makes sense. Things will break if there are two sanitizers that aren't mutually exclusive and both try to use the flag.

Reply:
@nikic  Could something like below be done so that sanitizers don't confuse with just "nosanitize" flag?

Add module flag "nosanitize" with a key that is enum SanitizerKind (with all supported sanitizers listed in it), defined in common header file for sanitizers. So, before running the pass, query the "nosanitize" flag and check that corresponding key value is present, to check if the pass has previously instrumented the IR.

M.addModuleFlag(Module::ModFlagBehavior::Override, "nosanitize", **SanitizerKind::Address**);

Reply:
You'd want a separate module flag for each sanitizer.

I'm a little skeptical that you pass pipeline is really doing what you want if you end up running sanitizer passes twice; that indicates you're running a bunch of other optimization passes twice without a good reason.

Reply:
Instrumenting a shadow load practically guarantees an immediate memory fault or other catastrophic behavior.  I assume any optimization passes that could cause such trouble are already idempotent.

Reply:
@nikic We don't have such instrumenting sanitizers yet, but having unique flags is LGTM

@efriedma-quic all previous proposals to add such "re-run" protection ended up with "the pipeline is wrong". But implementation is so trivial, so I don't mind just to have it implemented.